### PR TITLE
Hydrate secrets

### PR DESCRIPTION
**Pull request by LukasJenicek**
Issue: https://github.com/0xsequence/issue-tracker/issues/1823

### CODE REVIEW

**Comment by LukasJenicek** [comment]
image

File: cloudsecrets/hydrate.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+			return fmt.Errorf("passed config is nil")
+		}
+
+		v = v.Elem()
+	}
+
+	return processConfig(ctx, secretStorage, v)
+}
+
+func processConfig(ctx context.Context, storage SecretStorage, config reflect.Value) error {
```
Comment (VojtechVitek) [inline cloudsecrets/hydrate.go]: can we call this `hydrateStructFields()`?

Comment (VojtechVitek) [reply cloudsecrets/hydrate.go]: should we check for `v.Kind() == reflect.Struct` and error out otherwise?

File: cloudsecrets/hydrate.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	g, ctx := errgroup.WithContext(ctx)
+	var mux sync.Mutex
+
+	for i := 0; i < config.NumField(); i++ {
+		field := config.Field(i)
+		g.Go(func() error {
```
Comment (VojtechVitek) [inline cloudsecrets/hydrate.go]: I think we should create a goroutine only when fetching `SECRET:` value from GCP.

File: cloudsecrets/secret.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package cloudsecrets
+
+import "context"
+
+type SecretStorage interface {
+	// FetchSecret
+	// Abstraction to fetch value from secret store
+	// GCP uses versioning for each secret, you can always fetch the latest version by setting versionId="latest"
+	FetchSecret(ctx context.Context, secretId string, versionId string) (string, error)
```
Comment (VojtechVitek) [inline cloudsecrets/secret.go]: Do we need secret versioning? I'm thinking that maybe we support `:latest` only.

And if we'd need versioning in the future we'd create additional interface.

File: cloudsecrets/hydrate.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+				err := processConfig(ctx, storage, field)
+				mux.Unlock()
+				if err != nil {
+					return fmt.Errorf("failed to process config: %w", err)
+				}
+				return nil
+			}
+
+			if field.Kind() == reflect.String && field.CanSet() && strings.Contains(field.String(), "SECRET") {
+				_, secretName, found := strings.Cut(field.String(), ":")
```
Comment (VojtechVitek) [inline cloudsecrets/hydrate.go]: `strings.CutPrefix(s, "SECRET:")`

**Review by VojtechVitek — COMMENTED** [review]
Looks good. Few suggestions.

File: cloudsecrets/secret.go (hunk 3 above)
Comment (LukasJenicek) [reply cloudsecrets/secret.go]: I had it like this before.. I don't really know what's the best approach. 

I have no idea why GCP is versioning the secrets in the first place. Anyway I'll do it as you suggest

File: _examples/cmd/main.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	// fetch a projectId depends if you running project locally vs GKE
+	var projectId string
+	if metadata.OnGCE() {
+		projectId, err = metadata.ProjectID()
+		if err != nil {
+			log.Fatal("failed to get project ID from metadata:", err)
+		}
+	} else {
+		projectId = os.Getenv("GOOGLE_CLOUD_PROJECT")
+	}
```
Comment (david-littlefarmer) [inline _examples/cmd/main.go]:
```suggestion
	projectId = os.Getenv("GOOGLE_CLOUD_PROJECT")
	if metadata.OnGCE() {
		projectId, err = metadata.ProjectID()
		if err != nil {
			log.Fatal("failed to get project ID from metadata:", err)
		}
	}
```

File: cloudsecrets/hydrate.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+				return fmt.Errorf("failed to process config: %w", err)
+			}
+			continue
+		}
+
+		if field.Kind() == reflect.String && field.CanSet() && strings.Contains(field.String(), "SECRET") {
+			secretName, found := strings.CutPrefix(field.String(), "SECRET:")
+			if !found {
+				return fmt.Errorf("invalid config format: %s", field.String())
+			}
```
Comment (david-littlefarmer) [inline cloudsecrets/hydrate.go]:
```suggestion
		if field.Kind() == reflect.String && field.CanSet() && strings.HasPrefix(field.String(), "SECRET:") {
			secretName := strings.TrimPrefix(field.String(), "SECRET:")
```

File: cloudsecrets/hydrate.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+			g.Go(func() error {
+				secretValue, err := storage.FetchSecret(ctx, secretName)
+				if err != nil {
+					return err
```
Comment (david-littlefarmer) [inline cloudsecrets/hydrate.go]: error wrapping

**Review by david-littlefarmer — COMMENTED** [review]
Small things

File: cloudsecrets/hydrate.go (hunk 2 above)
Comment (LukasJenicek) [reply cloudsecrets/hydrate.go]: Yes. You are right. 

With this rewrite something is blocking spinning up more goroutines at the same times so I am looking into that at the moment. :eyes:

**Review by david-littlefarmer — APPROVED** [review]
LGTM

**Review by VojtechVitek — APPROVED** [review]
LGTM

**Pull request merged**